lib/cdo/throttle.rb

```diff
@@ -0,0 +1,45 @@
+require 'cdo/shared_cache'
+require 'dynamic_config/dcdo'
+
+module Cdo
+  module Throttle
+    CACHE_PREFIX = "cdo_throttle/".freeze
+
+    # @param [String] id - Unique identifier to throttle on.
+    # @param [Integer] limit - Number of requests allowed over period.
+    # @param [Integer] period - Period of time in seconds.
+    # @param [Integer] throttle_for - How long id should stay throttled in seconds. Optional.
+    # Defaults to Cdo::Throttle.throttle_time.
+    # @returns [Boolean] Whether or not the request should be throttled.
+    def self.throttle(id, limit, period, throttle_for = throttle_time)
+      full_key = CACHE_PREFIX + id.to_s
+      value = CDO.shared_cache.read(full_key) || empty_value
+      now = Time.now.utc
+      value[:request_timestamps] << now
+
+      if value[:throttled_until]&.future?
+        should_throttle = true
+      else
+        value[:throttled_until] = nil
+        earliest = now - period
+        value[:request_timestamps].select! {|timestamp| timestamp >= earliest}
```

Comment:
I'm slightly worried about this.  What kind of limit values do you think we will see in practice?  Throttling functions typically need to be really fast and this design grows linearly in time and space.

Reply:
yeah, this is the part that i feel weird about as well. the limit values are currently 100 requests / 60 seconds for identifiable users and 1000 requests / 60 seconds for unidentifiable users:
https://github.com/code-dot-org/code-dot-org/blob/21bb2de6058cba0713be866c244c50f1f82042cc/dashboard/app/controllers/profanity_controller.rb#L6-L9

the limits are configurable via DCDO values for each consumer. when a consumer is throttled, i am logging to honeybadger so we're notified and know if we need to adjust those limits

Reply:
do you have any thoughts for how to improve this? i'm trying to keep this array of timestamps small by evicting any values outside of the period (which is currently 60 seconds for all consumers) to mitigate, but there may be a better way to do this (or to track the requests differently to avoid this problem)

Reply:
One way to solve this is to track counts in buckets instead of the exact timestamps.  Each bucket would represent the number of requests that arrived during a particular window (say, 5 seconds) and you could sum up enough buckets to get the count for the interval that you want.  Old buckets would automatically age out of the interval as time progresses.  (Sorry if that's not super clear, this might be easier to explain with a sketch if it doesn't make sense.)